### PR TITLE
Adding mmu buffer yang files

### PR DESCRIPTION
**Pull request by AmitKaushik7**
#### Why I did it
Added Buffer yang files derived from sonic/mgmt-cvl/testdata/schema from sonic-mgmt-common,  used by mgmt-framework.

Updated BUFFER_PG|({ifname},)*|{pg_num} to BUFFER_PG|{ifname}|{pg_num} in sonic-buffer-pg.yang.
This change is required for configuration migration for dynamic port breakout operation.

Added sonic-buffer-queue.yang for BUFFER_QUEUE

Tables: BUFFER_POOL, BUFFER_PROFILE, BUFFER_PG, BUFFER_QUEUE.

#### How I did it
Defined Yang models for BUFFER tables based on Guideline doc:
https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md
and
https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md

#### How to verify it
sonic_yang_models package build.

#### Description for the changelog
MMU Buffer yang files (BUFFER_POOL/BUFFER_PROFILE/BUFFER_PG/BUFFER_QUEUE)

### CODE REVIEW

**Comment by AmitKaushik7** [comment]
Iguohan/Neethajohn   Could you review the PR and share you feedback

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:36]: need description for all features

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:26]: We want to keep this revision date or latest ?

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:38]: Plz remove the comments...

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:52]: If the file has tabs, plz convert to spaces.

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:42]: if logical interfaces are not allowed and this is always physical port, then better to use port_name\port. Agree?

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:50]: 0-7

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:40]: remove the commented line

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-queue.yang:49]: pattern must change 0-15,

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:45]: what is the unit, is it bytes, can you clarify?

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-queue.yang:38]: same for this comments, ...remove

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-queue.yang:47]: plz add descriptions for all leafs.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (dgsudharsan) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:49]: Please mandate the buffer pool

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:47]: should be mandatory ?

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (dgsudharsan) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:52]: Can you please confirm if we need to add buffer profile threshold mode?

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:41]: is there way to express the constraint that static_th and dynamic_th as mutually exclusive?

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:35]: We can have either static or dynamic, we may need choice statement here

**Review by praveen-li — COMMENTED** [review]
Updated comments from YANG WG.

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-queue.yang:58]: This PR has a dependency on https://github.com/Azure/sonic-swss/pull/1754, please merge this PR once the dependent PR is merged.

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:46]: xoff field is missing in the buffer pool table.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:55]: Please add descriptions for fields.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:40]: I think, the valid range is -6 to 3, please check and update.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:44]: Please refer the below SAI spec and update the default values for fields accordingly.

https://github.com/opencomputeproject/SAI/blob/master/inc/saibuffer.h

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:41]: We can try with count(field-name) = 0 or 1 condition in a must statement.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:18]: Update the revision.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:34]: As discussed in the meeting, can we add optional field mode (static/dynamic) as well?

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-pg.yang:50]: What if there is an overlap PG entries i.e same PG number for more than one entry, is that a valid config? if no, how are we planning to validate that? thru custom validation?

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-queue.yang:58]: Sure

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:34]: Venkat, there should not be a need to add an optional field "mode". Only one of static_th/dynamic_th must be specified and mode can be derived from threshold defined.  static_th/dynmic_th can be made mandatory and using choice  we can select one of then,

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:35]: Yes, will add a choice statement here.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:41]: I will try with choice statement.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:34]: This mode field is already present in the backend code and present in SAI spec as well under profile (SAI_BUFFER_PROFILE_ATTR_THRESHOLD_MODE), so, in order to match the existing implementation, we have to add it.

Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:34]: Please check bufferorch.cpp, the mode value is derived based on static_th/dynamic_th value and 2 SAI Attributes are set accordingly (SAI_BUFFER_PROFILE_ATTR_THRESHOLD_MODE 
 and SAI_BUFFER_PROFILE_ATTR_SHARED_DYNAMIC_TH/SAI_BUFFER_PROFILE_ATTR_SHARED_STATIC_TH).

Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: I wanted to have the mode under buffer-profile table because some platforms use it, but since the backend is assuming the mode from static_th/dynamic_th, we can avoid the mode field in the profile table and update platform-specific buffer defaults, thanks!

https://github.com/Azure/sonic-buildimage/blob/master/device/dell/x86_64-dellemc_s5232f_c3538-r0/DellEMC-S5232f-C32/buffers_defaults_t0.j2#L26

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pg.yang]: changed to port.

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-queue.yang]: changed to 0-15.

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pg.yang]: Removed Tabs

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pg.yang]: Removed comments.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: Note required. Backend can  derive it based on static_th/dynamic_th defined mode.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: Done.

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pool.yang]: Bytes. Updated the description for the same.

File: src/sonic-yang-models/yang-models/sonic-port-ingress-profile-list.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-port-ingress-profile-list.yang:34]: Where do we process this table in the backend?

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: Praveen, 
   I tried following choice/case statement for the threshold, but got the error mentioned below. 
Statement
=======
                choice threshold_type {
                    mandatory true;
                    case static {
                        leaf static_th {
                            type uint64;
                            description "The maximum size of the buffer pool the port can occupy";
                        }
                    }
                    case dynamic {
                        leaf dynamic_th {
                            type int32 {
                                range "-7..3";
                            }
                            description "The maximum proportion of the free size of the buffer pool the port can occupy (Alpha value)";
                        }
                    }
                }

Error:
====
           self.root = None
           self.sysLog(msg="Data Loading Failed:{}".format(str(e)), \
            debug=syslog.LOG_ERR, doPrint=True)
>          raise SonicYangException("Data Loading Failed\n{}".format(str(e)))
E          sonic_yang_ext.SonicYangException: Data Loading Failed
E          string indices must be integers

File: src/sonic-yang-models/yang-models/sonic-port-ingress-profile-list.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-port-ingress-profile-list.yang]: Renamed it to BUFFER_PORT_INGRESS_PROFILE_LIST, it is processed in bufferorch.cpp in the backend.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: Tried with choice but hitting error. Please refer comments from Praveen below.

File: src/sonic-yang-models/yang-models/sonic-buffer-pg.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pg.yang]: It should depend on the App code whether it needs to be overwritten or not. If we have 2 entries like "Ethernet4:3-4" and "Ethernet4:4-5", then both these configs are valid and should be applied in the order. Application in this case will be overriding  PG 4 with "Ethernet4:4-5" config. I am not aware if we can validate such overlap in yang.  The only solution I can think of is to split the PG(4 and 5) separately instead of range(say 4-5). This should be true not only for PG, but for other interfaces where we are specifying the range(Eg. Ethernet4-Ethernet10).

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-profile.yang]: Praveen, Do you see any issue with the above yang block  for choice.

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:28]: I'm not sure we allow any name here, I see the backend code written with a specific name e.g egress_lossless_pool, ingress_lossless_pool..etc, please check and restrict the name usage.

Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:28]: Hi Venkat
   egress_lossless_pool/ingress_lossy_pool etc are defined in buffer_defaults j2 files only. There is no restriction for the name in the cpp files and it is just a string and it just has to be unique only. I dont think there is a need of name restriction here. 
Thanks
Amit

Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:28]: Please see the code below, there is an assumption to use a particular pool name.
https://github.com/Azure/sonic-swss/blob/master/cfgmgr/buffermgr.cpp#L173

Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:28]: This is added only for ingress_lossless_pool in sonic-swss/cfgmgr/buffermgr.h.
#define INGRESS_LOSSLESS_PG_POOL_NAME "ingress_lossless_pool"
What about other buffer pools- egress/ingress lossy/lossless. (egress_lossy_pool/ingress_lossy_pool) etc?

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-queue.yang]: Venkat, Azure/sonic-swss#1754 is still not merged. Does the current PR has dependency on #1754.

**Comment by zhangyanzhao** [comment]
@venkatmahalingam will take a further look, then we can merge.

File: src/sonic-yang-models/yang-models/sonic-buffer-queue.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-queue.yang]: Yes, the current backend expectation is to have a table name (BUFFER_PROFILE) also while referencing to profile table from BUFFER_QUEUE and other tables.

127.0.0.1:6379[4]> hgetall BUFFER_QUEUE|Ethernet38|5-6
1) "profile"
2) "[BUFFER_PROFILE|egress_lossy_profile]"

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:28]: ok, let's have a string type for now to accept any pool name.

**Comment by venkatmahalingam** [comment]
> @venkatmahalingam will take a further look, then we can merge.

This PR has a dependency on Azure/sonic-swss#1754

File: src/sonic-yang-models/yang-models/sonic-buffer-pool.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-pool.yang:41]: Shouldn't it be mandatory as well?

File: src/sonic-yang-models/yang-models/sonic-buffer-port-egress-profile-list.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-port-egress-profile-list.yang:47]: Is it possible to check whether an egress profile is referenced here?
An egress profile is a buffer profile that references a buffer pool with `egress` as type

File: src/sonic-yang-models/yang-models/sonic-buffer-port-ingress-profile-list.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-port-ingress-profile-list.yang:47]: The same as egress profile list.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:51]: Per my understanding, headroom is on the ingress side only. But the buffer profiles are configured on both sides.
Better to change it to `reserved size`.

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:43]: I think the range should be the union of the legal range of all vendors 
`-8..7` should be better.
BTW the bounders are inclusive, right?

**Review by stephenxs — CHANGES_REQUESTED** [review]

File: src/sonic-yang-models/yang-models/sonic-buffer-profile.yang
Comment (stephenxs) [inline src/sonic-yang-models/yang-models/sonic-buffer-profile.yang:79]: There is another leaf `headroom_type` if the buffer model is dynamic. It is handled by dynamic buffer manager.

Check [schema](https://github.com/Azure/SONiC/blob/master/doc/qos/dynamically-headroom-calculation.md#schema-7) and [example](https://github.com/Azure/SONiC/blob/master/doc/qos/dynamically-headroom-calculation.md#dynamic-calculation-with-non-default-alpha) for details

**Comment by zhangyanzhao** [comment]
BRCM team will follow-up this.

File: src/sonic-yang-models/yang-models/sonic-buffer-port-egress-profile-list.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-port-egress-profile-list.yang:47]: No I dont think it can be checked for leafref.

File: src/sonic-yang-models/yang-models/sonic-buffer-port-ingress-profile-list.yang
Comment (AmitKaushik7) [reply src/sonic-yang-models/yang-models/sonic-buffer-port-ingress-profile-list.yang:47]: No I dont think it can be checked for leafref.

**Comment by AmitKaushik7** [comment]
@neethajohn  Can you see if this PR can be merged. Stephen had approved the changes. Builds & tests failure were not related to the changes in this PR. After rebase all builds are successful and tests are passing.

**Review by neethajohn — APPROVED** [review]

**Pull request merged**
